File: problem23.js
function isAbundant(number) {

  if ( number < 12 ) {
    return false;
  }    

  var sum = 0;
  var possibleFactor = 1;
    var sqrt = parseInt(Math.sqrt(number));
    while (possibleFactor <= sqrt) {
        if (number % possibleFactor == 0) {
            sum = sum + possibleFactor;

            var otherPossibleFactor = number / possibleFactor;
            if (otherPossibleFactor > possibleFactor && otherPossibleFactor != number ) {
                sum = sum + otherPossibleFactor;
            }
        }
        if ( sum > number ) {
          return true;
        }
        possibleFactor++;
    }

  if ( sum > number ) {
    return true;
  }

  return false;

}

function generateNAbundantNumber(n) {

  var abunNums = [12,18,20,24,30,36,40,42,48];
  for (var i = 50; i <=n; i++ ) {
    if ( isAbundant(i) ) {
      abunNums.push(i);
    }
  }
  return abunNums;

}

function sumOfNonAbundantNumbers(n) {
  
  var abunList = generateNAbundantNumber(n);
  
  var sum = 0;
  
  var isSumOfAbundant = false;
  var abundantNumber1 ;
  
  for ( var testNum = 1; testNum <= n; testNum++ ) {

    isSumOfAbundant = false;

    var j = abunList.length - 1;
    for ( var i = 0 ; abunList[i] < testNum; i++ ) {

      abundantNumber1 = abunList[i];  
      if ( abundantNumber1 > testNum ) {
        break;
      }

      var abundantNumber2 = abunList[j];
      while (( j > 0 ) && ( abundantNumber1 + abundantNumber2 > testNum )) {          
          abundantNumber2 = abunList[--j];
      }
        
      if ( testNum == abundantNumber1 + abundantNumber2 ) {
        //console.log("Identified one "+testNum+" "+abunList[i] +" "+ abunList[j]);
        isSumOfAbundant = true;
        break;
      }
    }
    
    if ( ! isSumOfAbundant ) {
      sum = sum + testNum;
      //console.log("Sum is "+sum+" for i "+i+" testNum "+testNum);
    }
    
  }

  return sum;

}

//sumOfNonAbundantNumbers(28123);
//console.log("Sum of Non Abundant numbers are "+sumOfNonAbundantNumbers(10000));

File: go/problem23.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(NonAbundantSum(10000))
}

func NonAbundantSum(limit int) int {

	var abundantNums []int
	for i := 3; i <= limit; i++ {
		if IsAbundant(i) {
			abundantNums = append(abundantNums, i)
		}
	}
	// fmt.Println(len(abundantNums))

	nonAbundantNumbersSum := 0

	isSumOfAbundant := false

	for testNum := 1; testNum <= limit; testNum++ {

		isSumOfAbundant = false

		j := len(abundantNums) - 1
		for i := 0; abundantNums[i] < testNum; i++ {

			abundantNumber1 := abundantNums[i]
			if abundantNumber1 > testNum {
				break
			}

			abundantNumber2 := abundantNums[j]
			for (j > 0) && (abundantNumber1+abundantNumber2 > testNum) {
				j = j - 1
				abundantNumber2 = abundantNums[j]
			}

			if testNum == abundantNumber1+abundantNumber2 {
				// fmt.Println("Identified one ", testNum, " ", abundantNums[i], " ", abundantNums[j])
				isSumOfAbundant = true
				break
			}
		}

		if !isSumOfAbundant {
			nonAbundantNumbersSum += testNum
		}

	}

	return nonAbundantNumbersSum

}

func IsAbundant(number int) bool {

	if number <= 0 {
		return false
	}

	if number == 1 || number == 2 {
		return false
	}

	var factorSum int = 1
	var i int
	var incrementor int = 1
	if number%2 != 0 {
		incrementor = 2
	} else {
		quotient := int(number) / 2
		if quotient != 2 {
			factorSum += quotient
		}
		factorSum += 2
	}

	for i = 3; i <= int(math.Sqrt(float64(number))); i += incrementor {
		if number%i == 0 {
			quotient := number / i
			if quotient != i {
				factorSum += quotient
			}
			factorSum += i
		}
	}

	if factorSum > number {
		return true
	} else {
		return false
	}
}
